File: tools/compose.py
#!/usr/bin/env python3
"""Compose a Palworld TCG card atlas for the Deck Maker template.

Layout must match the template's baked UVs: cells are filled row-major,
left-to-right, top-to-bottom, and each card fills its cell completely
(the template's own atlas has zero padding - verified by measurement).

  python3 compose.py --deck deck.json --art art/ --cards cards.json --out front.webp
"""
import argparse, json, os
from PIL import Image
from imgfix import solidify

COLS, ROWS = 10, 7          # must match the baked template's AtlasInfo GridSize
ROT = Image.ROTATE_270      # clockwise: landscape cards are turned right to read


def compose(deck, art_dir, landscape, size=8192):
    """deck: [{'code':..,'n':..}]  ->  RGBA atlas with one cell per physical card."""
    phys = [e["code"] for e in deck for _ in range(e["n"])]
    if len(phys) > COLS * ROWS:
        raise SystemExit(f"{len(phys)} cards exceeds the {COLS}x{ROWS} grid")
    atlas = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    for i, code in enumerate(phys):
        r, c = divmod(i, COLS)
        x0, y0 = round(c * size / COLS), round(r * size / ROWS)
        x1, y1 = round((c + 1) * size / COLS), round((r + 1) * size / ROWS)
        im = Image.open(os.path.join(art_dir, f"{code}.webp")).convert("RGBA")
        if code in landscape:
            im = im.transpose(ROT)
        # solidify BEFORE resizing: card art is matted against white, so the
        # antialiased edge pixels are light. Downscaling them straight would
        # leave a white rim that survives the material's alpha cutout.
        atlas.paste(solidify(im).resize((x1 - x0, y1 - y0), Image.LANCZOS), (x0, y0))
    return atlas, len(phys)


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--deck", required=True, help="JSON: [{code,n,name}]")
    p.add_argument("--art", required=True, help="directory of <CODE>.webp")
    p.add_argument("--cards", required=True, help="Palify /api/cards snapshot")
    p.add_argument("--out", required=True)
    p.add_argument("--size", type=int, default=8192, help="8192 keeps effect text readable")
    p.add_argument("--quality", type=int, default=95)
    a = p.parse_args()

    cards = json.load(open(a.cards))["cards"]
    landscape = {c["code"] for c in cards if c.get("landscape")}
    deck = json.load(open(a.deck))
    atlas, n = compose(deck, a.art, landscape, a.size)
    atlas.save(a.out, "WEBP", quality=a.quality, method=6, exact=True)
    print(f"{a.out}  {a.size}x{a.size}  cards={n}  {os.path.getsize(a.out)/1048576:.2f}MB")
    print(f"pass --cards {n} to patch.mjs so the deck is trimmed to match")


if __name__ == "__main__":
    main()

File: web/compose.js
// Browser port of tools/compose.py and tools/prepare_back.py.
//
// Layout must match the template's baked UVs: cells are filled row-major,
// left-to-right, top-to-bottom, and each card fills its cell completely (the
// template's own atlas has zero padding - verified by measurement).

import { solidify, toImageData, toCanvas } from './imgfix.js';

export const COLS = 10, ROWS = 7;   // must match the baked template's AtlasInfo GridSize
export const ROT = 90;    // clockwise, matching compose.py's ROTATE_270; verified against the atlas inside the approved v7 package
const CELL_ASPECT = 0.700;          // the template's cell; card art is 0.7156

const encode = (cv, quality) => cv.convertToBlob({ type: 'image/webp', quality });

/**
 * deck: [{code, n}] in deck order. loadArt(code) -> ImageBitmap.
 * landscape: Set of codes Palify marks landscape. Returns {blob, cards}.
 */
export async function composeAtlas(deck, loadArt, landscape, opts = {}) {
  const { size = 8192, quality = 0.95, onProgress } = opts;

  const physical = deck.flatMap(e => Array.from({ length: e.n }, () => e.code));
  if (physical.length > COLS * ROWS)
    throw new Error(`${physical.length} cards exceeds the ${COLS}x${ROWS} grid - bake a larger template`);

  const atlas = new OffscreenCanvas(size, size);
  const cx = atlas.getContext('2d');
  cx.imageSmoothingEnabled = true;
  cx.imageSmoothingQuality = 'high';

  // one decode+solidify per distinct code, reused across its copies
  const prepared = new Map();

  for (let i = 0; i < physical.length; i++) {
    const code = physical[i];
    if (!prepared.has(code)) {
      const bmp = await loadArt(code);
      // solidify BEFORE resizing: see imgfix.js
      prepared.set(code, toCanvas(solidify(toImageData(bmp, landscape.has(code) ? ROT : 0))));
      bmp.close?.();
    }
    const r = Math.floor(i / COLS), c = i % COLS;
    const x0 = Math.round(c * size / COLS), y0 = Math.round(r * size / ROWS);
    const x1 = Math.round((c + 1) * size / COLS), y1 = Math.round((r + 1) * size / ROWS);
    cx.drawImage(prepared.get(code), x0, y0, x1 - x0, y1 - y0);
    onProgress?.(i + 1, physical.length);
  }

  return { blob: await encode(atlas, quality), cards: physical.length, size };
}

/**
 * Resize a card back to the template's cell aspect (0.700). The card art is
 * 0.7156 but the cell is 0.700, so fronts are squashed 2.2%; the back is matched
 * to the same aspect so it doesn't sit at different proportions.
 */
export async function prepareBack(src, width = 1024, quality = 0.95) {
  const height = Math.round(width / CELL_ASPECT);
  const solid = toCanvas(solidify(toImageData(src)));
  const cv = new OffscreenCanvas(width, height);
  const cx = cv.getContext('2d');
  cx.imageSmoothingEnabled = true;
  cx.imageSmoothingQuality = 'high';
  cx.drawImage(solid, 0, 0, width, height);
  return { blob: await encode(cv, quality), width, height };
}
